src/wp-settings.php

```diff
@@ -412,6 +412,120 @@
+ *
+ * @since 6.5.0
+ *
+ * @global $_wp_for_core_plugins_compat_version
+ * @global $_found_incompatible_for_core_plugins
+ *
+ * @param string $plugin_absolute_path Absolute path to the plugin's file.
+ */
+$_is_plugin_compatible_with_wp = static function ( $plugin_absolute_path ) use ( $_wp_for_core_plugins_compat_version ) {
+	global $_found_incompatible_for_core_plugins;
+
+	static $plugins_dir_strlen;
+	if ( empty( $plugins_dir_strlen ) ) {
+		$plugins_dir_strlen = strlen( WP_PLUGIN_DIR . '/' );
+	}
+
+	$plugin = substr( $plugin_absolute_path, $plugins_dir_strlen );
+
+	// Not a for-core plugin.
+	$min_compat_version = _get_plugin_wp_min_compatible_version( $plugin );
+	if ( ! $min_compat_version ) {
+		return true;
+	}
+
+	// The plugin is already marked as incompatible.
+	if ( isset( $_found_incompatible_for_core_plugins[ $plugin ] ) ) {
+		return false;
+	}
+
+	// Get the Name and Version from the plugin's header.
+	$plugin_data = get_file_data(
+		$plugin_absolute_path,
+		array(
+			'Name'    => 'Plugin Name',
+			'Version' => 'Version',
+		),
+		'plugin'
+	);
+
+	// Whoops, something went wrong. Bail out.
+	if ( ! ( isset( $plugin_data['Version'] ) && isset( $plugin_data['Name'] ) ) ) {
+		return true;
+	}
+
+	// Plugin is compatible.
+	if ( version_compare( $plugin_data['Version'], $min_compat_version, '>=' ) ) {
+		return true;
+	}
+
+	// Found an incompatible for-core plugin. Add it to the found global for later batch processing.
+	$_found_incompatible_for_core_plugins[ $plugin ] = array(
+		'plugin_absolute_path' => $plugin_absolute_path,
+		'plugin_name'          => $plugin_data['Name'],
+		'version_deactivated'  => $plugin_data['Version'],
+		'version_compatible'   => $min_compat_version,
+	);
+
+	return false;
+};
+
+/**
+ * Handle incompatible for-core plugins.
+ *
+ * The handler does the following:
+ *     * Removes incompatible plugins from the "active_plugins" option.
+ *     * Stores incompatible plugins in "wp_force_deactivation_incompatible_plugins" option, which is used
+ *       in wp-admin to notify the user.
+ *
+ * @since 6.5.0
+ *
+ * @global $_found_incompatible_for_core_plugins
+ */
+$_handle_incompatible_for_core_plugins = static function () {
+	global $_found_incompatible_for_core_plugins;
+
+	$active_plugins           = (array) get_option( 'active_plugins', array() );
+	$active_plugins_by_plugin = array_flip( $active_plugins );
+
+	// Remove each of the found incompatible plugins from the "active_plugins" option.
+	foreach ( $_found_incompatible_for_core_plugins as $plugin_slug => $plugin ) {
+		unset( $active_plugins[ $active_plugins_by_plugin[ $plugin_slug ] ] );
+	}
+
+	// Update the 'active plugins' option, which no longer includes the incompatible plugins.
+	update_option( 'active_plugins', $active_plugins );
+
+	// Update the list of incompatible plugins to notify user in the admin.
+	update_option( 'wp_force_deactivation_incompatible_plugins', $_found_incompatible_for_core_plugins );
+};
```

Comment:
I've been concerned about the impacts to ALL sites of having this code run in these mission critical loader loops. Part of those concerns include:

* All sites will see a performance hit, even if they don't have the Gutenberg plugin activated.
* Writing to the database will also cause a performance hit, though only for those sites with an incompatible Gutenberg plugin version.
* How will it affect caching?

Today, Dion raised the following concerns https://core.trac.wordpress.org/ticket/60457 after the [Plugins Dependency was committed](https://core.trac.wordpress.org/changeset/57545): 

>Running DB Writes, particularly `update_option()` on the front-end, especially during the bootstrap of WordPress can have catastrophic consequences when thousands of requests all try to update the database and cache at the same time.
>
>This leads to a much higher chance of the Database and Cache becoming out-of-sync, which can cause snowballing issues. This happened on api.WordPress.org/plugins today, with WordPress getting stuck in a situation where it constantly tried to update the DB and failed (because the data had already been written there, but the cache said otherwise) and as a result never properly updated the cache.

Reply:
Instead of writing to `'wp_force_deactivation_incompatible_plugins'` option, what if:

Convert `$_found_incompatible_for_core_plugins` into a global that is allowed to leak out of `wp-settings.php` to be used by `_render_notice_for_deactivated_incompatible_plugins()` that is hooked into the `'admin_notices'` action?

Risks a future BC considerations as it will be publicly exposed.

Is there a way around the risk?

Yes, by passing the variable to a closure via `use` and having the closure hooked into `'admin_notices'` action.

See it in action https://3v4l.org/1BocJ.

BUT then its a static closure which is not (easily) unhookable for extender customization. 

Is that a problem?

Hmm ... Given this is a low-level Core internal handler, might be okay, though still not ideal for further customization needs.

Reply:
Commit https://github.com/WordPress/wordpress-develop/pull/5855/commits/d2ec763c671f4941d5bc41df82f45c6654b0892d removes the `'wp_force_deactivation_incompatible_plugins`' option and need to write to the database for it. It switches to use a lambda / anonymous function design while also avoiding global leaks.

Reply:
Nope that doesn't work as the value does not persist. Reverted https://github.com/WordPress/wordpress-develop/pull/5855/commits/37aff4c5d05467093fe4000adce540531fa22239.